pennylane/ops/qubit/special_unitary.py

```diff
@@ -0,0 +1,347 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the operation SpecialUnitary and
+its utility functions.
+"""
+# pylint: disable=arguments-differ
+from functools import lru_cache, reduce
+from itertools import product
+
+import numpy as np
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation
+
+_pauli_matrices = np.array(
+    [[[1, 0], [0, 1]], [[0, 1], [1, 0]], [[0, -1j], [1j, 0]], [[1, 0], [0, -1]]]
+)
+"""Single-qubit Paulis:    I                 X                   Y                  Z"""
+
+_pauli_letters = "IXYZ"
+"""Single-qubit Pauli letters that make up Pauli words."""
+
+
+@lru_cache
+def pauli_basis(num_wires):
+    r"""Compute all elements of the Pauli basis of the Lie algebra :math:`\mathfrak{su}(N)`
+    as a single, dense tensor.
+
+    Args:
+        num_wires (int): The number of wires on which the associated Pauli group acts.
+
+    Returns:
+        ndarray: All Pauli basis elements of :math:`\mathfrak{su}(N)`.
+
+    The basis has :math:`d=4^n-1` elements for :math:`n` qubits, yielding an output tensor
+    with shape ``(4**num_wires-1, 2**num_wires, 2**num_wires)`` and :math:`16^n-4^n` entries.
+    The identity Pauli word :math:`I^{\otimes n}` does not belong to :math:`\mathfrak{su}(N)`
+    and therefore is not included.
+
+    The basis elements are ordered (choose the description that suits you most)
+
+      - lexicographically.
+
+      - such that the term acting on the last qubit changes fastest, the one acting on the first
+        qubit changes slowest when iterating through the output.
+
+      - such that the basis index, written in base :math:`4`, contains the indices for the list
+        ``["I", "X", "Y", "Z"]``, in the order of the qubits
+
+      - such that for three qubits, the first basis elements are the Pauli words
+        ``"IIX", ""IIY", "IIZ", "IXI", "IXX", "IXY", "IXZ", "IYI"...``
+
+    .. note::
+
+        Note that this method internally handles a complex-valued tensor of size
+        ``(4**num_wires, 2**num_wires, 2**num_wires)``, which requires at least
+        ``4 ** (2 * num_wires - 13)`` GB of memory (at default precision).
+
+    **Example**
+
+    >>> pauli_basis(1)
+    array([[[ 0.+0.j,  1.+0.j],
+            [ 1.+0.j,  0.+0.j]],
+           [[ 0.+0.j, -0.-1.j],
+            [ 0.+1.j,  0.+0.j]],
+           [[ 1.+0.j,  0.+0.j],
+            [ 0.+0.j, -1.+0.j]]])
+
+    >>> pauli_basis(3).shape
+    (63, 8, 8)
+    """
+    if num_wires < 1:
+        raise ValueError(f"Require at least one wire, got {num_wires}.")
+    if num_wires > 7:
+        raise ValueError(
+            f"Creating the Pauli basis tensor for more than 7 wires (got {num_wires}) is deactivated."
+        )
+    return reduce(np.kron, (_pauli_matrices for _ in range(num_wires)))[1:]
+
+
+@lru_cache
+def pauli_words(num_wires):
```

Comment:
Just a suggestion, but could we add `basis` in the function name to make it clear that this function returns the basis?

Reply:
I now went for `pauli_basis_strings` and `pauli_basis_matrices` to be extra explicit about what they do. The word `word` is a bit ambiguous because the matrices also represent Pauli words...

Reply:
Much better, thanks!